File: app/src/main/java/com/djavid/bitcoinrate/view/adapter/TickerItem.java
package com.djavid.bitcoinrate.view.adapter;

import android.content.Context;
import android.support.v7.widget.LinearLayoutManager;
import android.util.Log;
import android.widget.ImageView;
import android.widget.TextView;

import com.annimon.stream.Collectors;
import com.annimon.stream.Stream;
import com.djavid.bitcoinrate.App;
import com.djavid.bitcoinrate.R;
import com.djavid.bitcoinrate.model.dto.LabelItemDto;
import com.djavid.bitcoinrate.model.dto.heroku.Subscribe;
import com.djavid.bitcoinrate.model.dto.heroku.Ticker;
import com.djavid.bitcoinrate.util.Codes;
import com.djavid.bitcoinrate.util.DateFormatter;
import com.mindorks.placeholderview.PlaceHolderView;
import com.mindorks.placeholderview.annotations.Layout;
import com.mindorks.placeholderview.annotations.NonReusable;
import com.mindorks.placeholderview.annotations.Resolve;
import com.mindorks.placeholderview.annotations.View;

import java.util.ArrayList;
import java.util.List;

import io.reactivex.subjects.PublishSubject;
import io.reactivex.subjects.Subject;


@NonReusable
@Layout(R.layout.recycler_item_ticker)
public class TickerItem {

    @View(R.id.tv_ticker_title)
    private TextView tv_ticker_title;
    @View(R.id.tv_price_change)
    private TextView tv_price_change;
    @View(R.id.tv_market_cap)
    private TextView tv_market_cap;
    @View(R.id.iv_ticker_icon)
    private ImageView iv_ticker_icon;
    @View(R.id.tickerValue)
    private TextView tickerValue;
    @View(R.id.label_container)
    private PlaceHolderView label_container;
    @View(R.id.v_divider2)
    private android.view.View v_divider2;

    private Ticker tickerItem;
    private List<LabelItemDto> labels;

    private final String TAG = this.getClass().getSimpleName();
    private Context mContext;
    private PlaceHolderView mPlaceHolderView;
    private String tv_price;
    private int price_change_color;
    private String price_change;
    Subject<String> mObservable = PublishSubject.create();


    public TickerItem(Context mContext, PlaceHolderView mPlaceHolderView, Ticker tickerItem,
                      List<Subscribe> subscribes) {
        this.mContext = mContext;
        this.mPlaceHolderView = mPlaceHolderView;
        this.tickerItem = tickerItem;

        labels = new ArrayList<>();
        for (Subscribe item : subscribes) {
            labels.add(new LabelItemDto(item.getId(), item.getValue(),
                    item.isTrendingUp(), item.getChange_percent()));
        }
        labels = sortLabels(labels);
    }

    public TickerItem(Context mContext, PlaceHolderView mPlaceHolderView, Ticker tickerItem) {
        this.tickerItem = tickerItem;
        this.mContext = mContext;
        this.mPlaceHolderView = mPlaceHolderView;
        labels = new ArrayList<>();
    }

    @Resolve
    private void onResolved() {

        try {

            String title = Codes.getCryptoCurrencyId(tickerItem.getCryptoId());
            if (title.length() > 12) {
                tv_ticker_title.setTextSize(13f);
            }
            tv_ticker_title.setText(title);

            tv_market_cap.setText(DateFormatter.convertMarketCap(tickerItem.getTicker().getMarket_cap_usd()));

            iv_ticker_icon.setImageResource(Codes.getCurrencyImage(tickerItem.getCryptoId()));
            tickerValue.setText(tv_price);
            tv_price_change.setTextColor(price_change_color);
            tv_price_change.setText(price_change);

            LinearLayoutManager layoutManager =
                    new LinearLayoutManager(mContext, LinearLayoutManager.HORIZONTAL, false);
            label_container.setLayoutManager(layoutManager);


            for (LabelItemDto item : labels) {
                label_container.addView(new LabelItem(mContext, label_container, item, this));
            }
            label_container.addView(new LabelItem(mContext, label_container, new LabelItemDto(), this));

        } catch (Exception e) {
            e.printStackTrace();
        }

    }

    private List<LabelItemDto> sortLabels(List<LabelItemDto> subscribes) {

        List<LabelItemDto> sorted_subscribes;

        sorted_subscribes = Stream.of(subscribes)
                .sorted((a, b) -> {

                    if (a.getChange_percent() == 0 && b.getChange_percent() == 0) {

                        return a.getValue().compareTo(b.getValue());

                    } else if (a.getChange_percent() != 0 && b.getChange_percent() != 0) {

                        return Double.compare(a.getChange_percent(), b.getChange_percent());

                    } else if (a.getChange_percent() == 0 && b.getChange_percent() != 0) {

                        return 1;

                    } else if (a.getChange_percent() != 0 && b.getChange_percent() == 0) {

                        return -1;

                    } else {

                        return 0;

                    }

                })
                .collect(Collectors.toList());

        return sorted_subscribes;
    }


    public Ticker getTickerItem() {
        return tickerItem;
    }
    public List<LabelItemDto> getLabels() {
        return labels;
    }


    public void addLabelItem(LabelItemDto new_item) {

        labels.add(new_item);

        label_container.removeAllViews(); //is it needed?
        for (LabelItemDto item : labels) {
            label_container.addView(new LabelItem(mContext, label_container, item, this));
        }
        label_container.addView(new LabelItem(mContext, label_container, new LabelItemDto(), this));

    }

    void deleteLabel(LabelItem labelItem) {
        labels.remove(labelItem.labelItemDto);
        label_container.removeView(labelItem);
    }

    public void setPrice(String price) {
        if (tickerItem != null)
            tv_price = price + " " + tickerItem.getCountryId();

        if (tickerValue != null) {
            tickerValue.setText(tv_price);
        }
    }

    public void setPriceChange(double change) {

        Log.i(TAG, "setPriceChange(" + change + ")");

        if (change > 0) {
            price_change = "+" + change + "%";
            price_change_color = App.getContext().getResources().getColor(R.color.colorPriceChangePos);
        } else {
            price_change = change + "%";
            price_change_color = App.getContext().getResources().getColor(R.color.colorPriceChangeNeg);
        }

        if (tv_price_change != null) {
            tv_price_change.setTextColor(price_change_color);
            tv_price_change.setText(price_change);
        }
    }

}

File: app/src/main/kotlin/com/djavid/bitcoinrate/view/ticker/TickerItem.kt
package com.djavid.bitcoinrate.view.ticker

import android.content.Context
import android.util.Log
import android.widget.ImageView
import android.widget.TextView
import com.bumptech.glide.Glide
import com.djavid.bitcoinrate.R
import com.djavid.bitcoinrate.app.App
import com.djavid.bitcoinrate.model.heroku.Subscribe
import com.djavid.bitcoinrate.model.heroku.Ticker
import com.djavid.bitcoinrate.model.project.LabelItemDto
import com.djavid.bitcoinrate.util.Codes
import com.djavid.bitcoinrate.util.PriceConverter
import com.mindorks.placeholderview.PlaceHolderView
import com.mindorks.placeholderview.annotations.Layout
import com.mindorks.placeholderview.annotations.NonReusable
import com.mindorks.placeholderview.annotations.Resolve
import com.mindorks.placeholderview.annotations.View
import io.reactivex.subjects.PublishSubject
import io.reactivex.subjects.Subject
import java.util.*


@NonReusable
@Layout(R.layout.recycler_item_ticker)
class TickerItem {

    @View(R.id.tv_ticker_title)
    private val tv_ticker_title: TextView? = null
    @View(R.id.tv_price_change)
    private val tv_price_change: TextView? = null
    @View(R.id.tv_market_cap)
    private val tv_market_cap: TextView? = null
    @View(R.id.iv_ticker_icon)
    private val iv_ticker_icon: ImageView? = null
    @View(R.id.tickerValue)
    private val tickerValue: TextView? = null
    @View(R.id.label_container)
    private val label_container: PlaceHolderView? = null
    @View(R.id.v_divider2)
    private val v_divider2: android.view.View? = null

    var tickerItem: Ticker? = null
        private set
    private var labels: MutableList<LabelItemDto>? = null

    private val TAG = this.javaClass.simpleName
    private var mContext: Context? = null
    private var mPlaceHolderView: PlaceHolderView? = null
    private var tv_price: String? = null
    private var price_change_color: Int = 0
    private var price_change: String? = null
    internal var mObservable: Subject<String> = PublishSubject.create()


    constructor(mContext: Context, mPlaceHolderView: PlaceHolderView, tickerItem: Ticker,
                subscribes: List<Subscribe>) {
        this.mContext = mContext
        this.mPlaceHolderView = mPlaceHolderView
        this.tickerItem = tickerItem

        labels = ArrayList()
        for (item in subscribes) {
            labels!!.add(LabelItemDto(item.id!!, item.value,
                    item.isTrendingUp, item.change_percent))
        }
        labels = sortLabels(labels)
    }

    constructor(mContext: Context, mPlaceHolderView: PlaceHolderView, tickerItem: Ticker) {
        this.tickerItem = tickerItem
        this.mContext = mContext
        this.mPlaceHolderView = mPlaceHolderView
        labels = ArrayList()
    }

    @Resolve
    private fun onResolved() {

        try {

            val title = Codes.getCryptoCurrencyId(tickerItem!!.cryptoId)
            if (title.length > 12) {
                tv_ticker_title!!.textSize = 13f
            }
            tv_ticker_title!!.text = title

            tv_market_cap!!.text = PriceConverter.convertMarketCap(tickerItem!!.ticker!!.market_cap_usd)

            println(tickerItem!!.cryptoId)
            val image_url = Codes.getCryptoCoinImage(tickerItem!!.cryptoId)
            Glide.with(mContext!!)
                    .asBitmap()
                    .load(image_url)
                    .into(iv_ticker_icon!!)

            tickerValue!!.text = tv_price
            tv_price_change!!.setTextColor(price_change_color)
            tv_price_change.text = price_change

            val layoutManager = LinearLayoutManager(mContext, LinearLayoutManager.HORIZONTAL, false)
            label_container!!.layoutManager = layoutManager


            for (item in labels!!) {
                label_container.addView(LabelItem(mContext, label_container, item, this))
            }
            label_container.addView(LabelItem(mContext, label_container, LabelItemDto(), this))

        } catch (e: Exception) {
            e.printStackTrace()
        }

    }

    private fun sortLabels(subscribes: List<LabelItemDto>): MutableList<LabelItemDto> {

        val sorted_subscribes: List<LabelItemDto>

//        sorted_subscribes = Stream.of(subscribes)
//                .sorted { a, b ->
//
//                    if (a.change_percent == 0.0 && b.change_percent == 0.0) {
//
//                        return@Stream.of(subscribes)
//                                .sorted a . value !!. compareTo b.value!!
//
//                    } else if (a.change_percent != 0.0 && b.change_percent != 0.0) {
//
//                        return@Stream.of(subscribes)
//                                .sorted java . lang . Double . compare a.change_percent, b.change_percent)
//
//                    } else if (a.change_percent == 0.0 && b.change_percent != 0.0) {
//
//                        return@Stream.of(subscribes)
//                                .sorted 1
//
//                    } else if (a.change_percent != 0.0 && b.change_percent == 0.0) {
//
//                        return@Stream.of(subscribes)
//                                .sorted - 1
//
//                    } else {
//
//                        return@Stream.of(subscribes)
//                                .sorted 0
//
//                    }
//
//                }
//                .collect<List<LabelItemDto>, Any>(Collectors.toList())
	
		return emptyList<LabelItemDto>().toMutableList()
    }

    fun getLabels(): List<LabelItemDto>? {
        return labels
    }


    fun addLabelItem(new_item: LabelItemDto) {

        labels!!.add(new_item)

        label_container!!.removeAllViews() //is it needed?
        for (item in labels!!) {
            label_container.addView(LabelItem(mContext, label_container, item, this))
        }
        label_container.addView(LabelItem(mContext, label_container, LabelItemDto(), this))

    }

    internal fun deleteLabel(labelItem: LabelItem) {
        labels!!.remove(labelItem.labelItemDto)
        label_container!!.removeView(labelItem)
    }

    fun setPrice(price: String) {
        if (tickerItem != null)
            tv_price = price + " " + tickerItem!!.countryId

        if (tickerValue != null) {
            tickerValue.text = tv_price
        }
    }

    fun setPriceChange(change: Double) {

        Log.i(TAG, "setPriceChange($change)")

        if (change > 0) {
            price_change = "+$change%"
            price_change_color = App.context.resources.getColor(R.color.colorPriceChangePos)
        } else {
            price_change = "$change%"
            price_change_color = App.context.resources.getColor(R.color.colorPriceChangeNeg)
        }

        if (tv_price_change != null) {
            tv_price_change.setTextColor(price_change_color)
            tv_price_change.text = price_change
        }
    }

}
